### PR TITLE
[CELEBORN-1490][CIP-6] Introduce tier factory and master agent in flink hybrid shuffle

### PR DESCRIPTION
**Pull request by reswqa**
### What changes were proposed in this pull request?

Implement `TieredFactory` interface to adapt Flink hybrid shuffle and also introduce `CelebornTierMasterAgent`.

Note: Only the last commit need review in this PR.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
No need.

### CODE REVIEW

File: client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierFactory.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  @Override
+  public void setup(Configuration configuration) {
+    conf = FlinkUtils.toCelebornConf(configuration);
+    this.bufferSizeBytes = ConfigurationParserUtils.getPageSize(configuration);
+  }
+
+  @Override
+  public TieredStorageMemorySpec getMasterAgentMemorySpec() {
+    return new TieredStorageMemorySpec(getCelebornTierName(), 0);
```
[the remark this answers is not in the document]
Comment (reswqa) [reply client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierFactory.java:71]: Yes, master and consumer both doesn't need guaranteed buffer(Used to tell flink the minimum buffer that must be reserved in the buffer pool).

File: client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      executor.execute(
+          () -> {
+            try {
+              for (Integer shuffleId : shuffleIds) {
+                lifecycleManager.unregisterShuffle(shuffleId);
+                shuffleTaskInfo.removeExpiredShuffle(shuffleId);
+              }
+              shuffleResourceTracker.unRegisterJob(jobID);
+            } catch (Throwable throwable) {
+              LOG.error("Encounter an error when unregistering job: {}.", jobID, throwable);
```
[the remark this answers is not in the document]
Comment (reswqa) [reply client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java:122]: I checked this part of the logic again, and there should be no chance to raise, so I removed the try-catch statement now.

File: client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    FlinkResultPartitionInfo resultPartitionInfo =
+        new FlinkResultPartitionInfo(jobID, resultPartitionID);
+    ShuffleResourceDescriptor shuffleResourceDescriptor =
+        shuffleTaskInfo.genShuffleResourceDescriptor(
+            resultPartitionInfo.getShuffleId(),
+            resultPartitionInfo.getTaskId(),
+            resultPartitionInfo.getAttemptId());
+    Set<Integer> shuffleIds = jobShuffleIds.get(jobID);
+    if (shuffleIds == null) {
+      throw new RuntimeException("Can not find job in master agent, job: " + jobID);
```
[the remark this answers is not in the document]
Comment (reswqa) [reply client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java:140]: You can think of this as a pure sanity check, and not expected to happen.

This exception is not thrown to the user but it is handled by the flink framework, which triggers the job failover.

**Review by SteNicholas — APPROVED** [review]

**Review by RexXiong — APPROVED** [review]
LGTM

**Pull request closed without merging**